src/main/java/org/apache/sysml/api/mlcontext/MLContextConversionUtil.java

```diff
@@ -361,13 +363,57 @@ public static MatrixObject dataFrameToMatrixObject(String variableName, DataFram
 
 		JavaRDD<Row> javaRDD = dataFrame.javaRDD();
 		JavaPairRDD<Row, Long> prepinput = javaRDD.zipWithIndex();
-		JavaPairRDD<MatrixIndexes, MatrixBlock> out = prepinput.mapPartitionsToPair(new DataFrameToBinaryBlockFunction(
-				matrixCharacteristics, isVectorBasedDataFrame));
+		JavaPairRDD<MatrixIndexes, MatrixBlock> out = prepinput
+				.mapPartitionsToPair(new DataFrameToBinaryBlockFunction(matrixCharacteristics, isVectorBasedDataFrame));
 		out = RDDAggregateUtils.mergeByKey(out);
 		return out;
 	}
 
 	/**
+	 * If the MatrixFormat of the DataFrame has not been explicitly specified,
+	 * attempt to determine the proper MatrixFormat.
+	 * 
+	 * @param dataFrame
+	 *            the Spark {@code DataFrame}
+	 * @param matrixMetadata
+	 *            the matrix metadata, if available
+	 */
+	public static void determineMatrixFormatIfNeeded(DataFrame dataFrame, MatrixMetadata matrixMetadata) {
+		MatrixFormat matrixFormat = matrixMetadata.getMatrixFormat();
+		if (matrixFormat != null) {
+			return;
+		}
+		StructType schema = dataFrame.schema();
+		boolean hasID = false;
+		try {
+			schema.fieldIndex("ID");
+			hasID = true;
+		} catch (IllegalArgumentException iae) {
+		}
+		Row firstRow = dataFrame.first();
+		MatrixFormat mf = null;
+		if (hasID) {
+			Object object = firstRow.get(1);
```

Comment:
I would stick with this implementation for now, but at some point if this happens to become a bottleneck, we should explore determining the type directly from the schema.